### PR TITLE
Feature: Add `FixedSizeListBuilder`

### PR DESCRIPTION
**Pull request by connortsui20**
Tracking Issue: https://github.com/vortex-data/vortex/issues/4372

Adds support for `FixedSizeListBuilder: ArrayBuilder`. I don't actually _use_ the builder where it needs to be yet, but I am going to leave that for a followup PR.

Additionally adds unit tests for the builder, making sure to test the degenerate cases.

Also note that some of the TODOs with respect to the `append_zero` and `append_null`s will be solved in the next PR I do so don't worry too much about those...

### CODE REVIEW

File: vortex-array/src/builders/primitive.rs [hunk 1]
```diff
@@ -141,6 +141,7 @@ impl<T: NativePType> ArrayBuilder for PrimitiveBuilder<T> {
         self.nulls.append_n_non_nulls(n);
     }
 
+    // TODO(connor): Shouldn't this check if the dtype is nullable? Otherwise this is just wrong...
```
Comment (joseph-isaacs) [inline vortex-array/src/builders/primitive.rs]: this is checked in the outer call?

Comment (connortsui20) [reply vortex-array/src/builders/primitive.rs]: we **hope** that it is checked in the outer call...

File: vortex-array/src/builders/fixed_size_list.rs [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+    fn append_nulls(&mut self, n: usize) {
+        let element_count = n * self.list_size() as usize;
+
+        // TODO(connor): Does the behavior of `append_nulls` make sense here?
+        // Use nulls if the element type is nullable, otherwise use zeros.
+        if self.element_dtype().is_nullable() {
+            self.elements_builder.append_nulls(element_count);
+        } else {
+            self.elements_builder.append_zeros(element_count);
+        }
```
Comment (joseph-isaacs) [inline vortex-array/src/builders/fixed_size_list.rs:171]: todo: append_default?

File: vortex-array/src/builders/fixed_size_list.rs [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+
+    fn finish(&mut self) -> ArrayRef {
+        let final_len = self.len();
+        assert_eq!(
+            self.elements_builder.len(),
+            final_len * self.list_size() as usize,
+            "elements length must be equal to the array length times the list size"
+        );
+
+        // TODO(connor): Use `new_unchecked` here.
```
Comment (joseph-isaacs) [inline vortex-array/src/builders/fixed_size_list.rs:214]: why wait?

File: vortex-array/src/builders/list.rs [hunk 4]
```diff
@@ -22,6 +22,7 @@ pub struct ListBuilder<O: NativePType> {
     /// Represents the offsets into the values array.
     index_builder: PrimitiveBuilder<O>,
     nulls: LazyNullBufferBuilder,
+    // TODO(connor): This can probably be removed since we store it in the `dtype` field as well.
     nullability: Nullability,
```
Comment (joseph-isaacs) [inline vortex-array/src/builders/list.rs:26]: remove?

File: vortex-array/src/builders/fixed_size_list.rs (hunk 3 above)
Comment (connortsui20) [reply vortex-array/src/builders/fixed_size_list.rs:214]: I'm going to fix this all at once in a separate PR (see #4420)

File: vortex-array/src/builders/list.rs (hunk 4 above)
Comment (connortsui20) [reply vortex-array/src/builders/list.rs:26]: I'm going to do this in a followup PR where I clean up the entire `builders` module

File: vortex-array/src/builders/primitive.rs (hunk 1 above)
Comment (connortsui20) [reply vortex-array/src/builders/primitive.rs]: Am going to fix this in a followup PR

File: vortex-array/src/arrays/fixed_size_list/tests/nested.rs [hunk 5]
```diff
@@ -153,6 +208,36 @@ fn test_deeply_nested_fsl() {
                     )
             )
     ));
+
```
Comment (joseph-isaacs) [inline vortex-array/src/arrays/fixed_size_list/tests/nested.rs:211]: do we need to test null in deeply nested lists?

**Review by joseph-isaacs — APPROVED** [review]

File: vortex-array/src/arrays/fixed_size_list/tests/nested.rs (hunk 5 above)
Comment (connortsui20) [reply vortex-array/src/arrays/fixed_size_list/tests/nested.rs:211]: Yes, but also I just tagged this change onto this PR even though it has only has a bit of relevance. Once I implement the rest of the functionality needed for fixed-size lists, I will definitely be heavily testing nulls in deeply nested fsls in order to test tensors
